File: 2019/rs/src/day4.rs
pub fn day4(input_path: &str) {
    let min_max: Vec<i32> = std::fs::read_to_string(input_path)
        .unwrap_or_else(|e| panic!("Failed to open {}: {}", input_path, e))
        .trim()
        .split("-")
        .map(|x| {
            x.parse::<i32>()
                .unwrap_or_else(|e| panic!("Failed to parse {}", e))
        })
        .collect();
    assert_eq!(min_max.len(), 2);
    let min = min_max[0];
    let max = min_max[1];

    let mut part1 = 0;
    let mut part2 = 0;
    let mut x = min;
    while x <= max {
        let digits = x
            .to_string()
            .chars()
            .map(|d| d.to_digit(10).unwrap())
            .collect::<Vec<u32>>();
        x += 1;

        if !is_non_decreasing(&digits[..]) {
            continue;
        }
        if has_same_adjacent_digits(&digits[..]) {
            part1 += 1;
            if has_same_adjacent_pair(&digits[..]) {
                part2 += 1;
            }
        }
    }
    assert_eq!(part1, 1955);
    assert_eq!(part2, 1319);
}

fn is_non_decreasing(digits: &[u32]) -> bool {
    let mut maybe_last = None::<u32>;
    for x in digits.iter() {
        if let Some(last) = maybe_last {
            if last > *x {
                return false;
            }
        }
        maybe_last = Some(*x);
    }
    true
}

fn has_same_adjacent_digits(digits: &[u32]) -> bool {
    let mut maybe_last = None::<u32>;
    for x in digits.iter() {
        if let Some(last) = maybe_last {
            if last == *x {
                return true;
            }
        }
        maybe_last = Some(*x);
    }
    false
}

fn has_same_adjacent_pair(mut digits: &[u32]) -> bool {
    while !digits.is_empty() {
        if let [_] = digits {
            return false;
        }
        if let [x, y] = digits {
            return x == y;
        }
        if let [x, y, z] = digits[..3] {
            if x == y {
                if x != z {
                    return true;
                }
                let mut i = 3;
                while i < digits.len() {
                    if digits[i] != x {
                        break;
                    }
                    i += 1;
                }
                digits = &digits[i..];
            } else {
                digits = &digits[1..];
            }
        } else {
            panic!("Unreachable");
        }
    }
    false
}

File: 2019/py/day4.py
#!/usr/bin/env python3

import itertools


def day4(input_path):
    with open(input_path) as f:
        min_val, max_val = map(int, f.readline().split("-"))

    part1 = 0
    part2 = 0
    for x in range(min_val, max_val + 1):
        x = str(x)
        if not is_non_decreasing(x):
            continue
        if has_repeats(x):
            part1 += 1
            if has_exact_repeats(x):
                part2 += 1
    assert part1 == 1955
    assert part2 == 1319


def is_non_decreasing(x):
    i1, i2 = itertools.tee(x)
    next(i2, None)
    for a, b in zip(i1, i2):
        if a > b:
            return False
    return True


def has_repeats(x):
    for i in range(len(x) - 1):
        if x[i] == x[i + 1]:
            return True
    return False


def has_exact_repeats(x):
    i = 0
    while i < len(x) - 1:
        val = x[i]
        if val != x[i + 1]:
            i += 1
            continue
        if i + 1 == len(x) - 1 or x[i + 2] != val:
            return True
        while i < len(x) and x[i] == val:
            i += 1
    return False
